File: velocity-position-rotation_mode/drive_modes/Lift_Robot.py
class Lift_Robot:

    def __init__(self, robot):
        self.robot = robot

    def run(self):

        #Driving main wheels
        speed = self.robot.joystick.getRawAxis(5)
        self.robot.fl_motor.set(speed)
        self.robot.bl_motor.set(speed)
        self.robot.fr_motor.set(speed)
        self.robot.br_motor.set(speed)

        back_lift_pos = self.robot.back_lift.getSelectedSensorPosition() 
            

            
      
        #setting lift power
        lift_power = -self.robot.navx.getPitch() / (self.robot.lift_divider *1.5)

        print("back lift pos:", back_lift_pos)

        #back lift wheel control
        if self.robot.joystick.getPOV(0) == 0:
            self.robot.back_lift_wheel.set(-1)
        elif self.robot.joystick.getPOV(0) == 180:
            self.robot.back_lift_wheel.set(1)
        else:
          self.robot.back_lift_wheel.set(0)

        #when up (a) button is pressed (lift goes up)
        if self.robot.joystick.getRawButton(1):
            #setting back lift
            if back_lift_pos < -72000:
                self.robot.back_lift.set(self.robot.back_lift_speed_up)
            else:
                self.robot.back_lift.set(0)

            #setting front lift
            if lift_power < 0:
                lift_power = lift_power * self.robot.lift_speed_up
            else:
                lift_power = lift_power * -self.robot.lift_speed_down
                

            print(f"lift power: {lift_power}")       
            self.robot.front_lift.set(lift_power)

        #when down (y) button is pressed (lift goes down)
        elif self.robot.joystick.getRawButton(4):
            #setting back lift
            if back_lift_pos > -1780000:
                self.robot.back_lift.set(-self.robot.back_lift_speed_down)
            else:
                self.robot.back_lift.set(0)

            #setting front lift
            if lift_power < 0:
                lift_power = lift_power * self.robot.lift_speed_up
            else:
                lift_power = lift_power * -self.robot.lift_speed_down
            

            print(f"lift power: {lift_power}")           
            self.robot.front_lift.set(lift_power)
        
        else:
            # if nothing is pressed
            self.robot.back_lift.set(0)
            #self.robot.back_lift.set(-.1) #-self.robot.holding_back_lift)
            self.robot.front_lift.set(-.2) #-self.robot.holding_front_lift)

        #transition code
        if self.robot.climb_toggle:
            self.robot.front_lift_heights_index = 0
            return 'lift_robot'
        self.robot.front_lift.setQuadraturePosition(0, 0)
        return 'velocity'

File: kotlin/src/main/java/frc/robot/drivemodes/LiftRobot.kt
package frc.robot.drivemodes.liftrobot

import frc.robot.*

class Lift_Robot(robot : Robot) : State {

  private val m_robot = robot

  override public fun run() : String {
    // Driving main wheels
    val speed = Devices.joystick.getRawAxis(5)
    Devices.flMotor.set(speed)
    Devices.blMotor.set(speed)
    Devices.frMotor.set(speed)
    Devices.brMotor.set(speed)

    val backLiftPos = Devices.backLift.getSelectedSensorPosition() 

    // setting lift power
    var liftPower = -Devices.navx.getPitch() / (m_robot.liftDivider *1.5)

    // back lift wheel control
    if (Devices.joystick.getPOV(0) == 0)
      Devices.backLiftWheel.set(-1.0)
    else if (Devices.joystick.getPOV(0) == 180)
      Devices.backLiftWheel.set(1.0)
    else
      Devices.backLiftWheel.set(0.0)

    // when up (a) button is pressed (lift goes up)
    if (Devices.joystick.getRawButton(1)) {
      // setting back lift
      if (backLiftPos < -72000)
        Devices.backLift.set(m_robot.backLiftSpeedUp)
      else
        Devices.backLift.set(0.0)

      // setting front lift
      if (liftPower < 0)
          liftPower = liftPower * m_robot.liftSpeedUp
      else
          liftPower = liftPower * -m_robot.liftSpeedDown
          
      Devices.frontLift.set(liftPower)
    }

    // when down (y) button is pressed (lift goes down)
    else if (Devices.joystick.getRawButton(4)) {
      // setting back lift
      if (backLiftPos > -1780000)
        Devices.backLift.set(-m_robot.backLiftSpeedDown)
      else
        Devices.backLift.set(0.0)

      // setting front lift
      if (liftPower < 0)
          liftPower = liftPower * m_robot.liftSpeedUp
      else
          liftPower = liftPower * -m_robot.liftSpeedDown
      
      Devices.frontLift.set(liftPower)
    }
    else {
      // if nothing is pressed
      Devices.backLift.set(0.0)
      Devices.backLift.set(-.1)  // -self.robot.holding_back_lift)
      Devices.frontLift.set(-.2) // -self.robot.holding_front_lift)
    }

    // transition code
    if (m_robot.climbToggle) {
        m_robot.frontLiftHeightsIndex = 0
        return "lift_robot"
    }
    Devices.frontLift.setSelectedSensorPosition(0, 0, TIMEOUT_MS)
    return "velocity"
  }
}
